### PR TITLE
feat: open catalogs and schemas in parallel

### LINKED ISSUE #1517: Open catalogs and schemas in parallel
### What type of enhancement is this?

Performance

### What does the enhancement do?

In catalog manager(especially remote), it initialize the catalogs and schemas serially. 

Maybe we can initialize the catalogs and schemas in parallel too, just like tables.

https://github.com/GreptimeTeam/greptimedb/blob/c8301feed7310937900d03d6f9bd7649f625cc9e/src/catalog/src/remote/manager.rs#L233

### Implementation challenges

_No response_

### PR DESCRIPTION
**Pull request by gitccl**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?
Initialize the catalogs and schemas in parallel.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
close #1517

### CODE REVIEW

File: src/catalog/src/remote/manager.rs [hunk 1]
```diff
@@ -332,6 +201,203 @@ impl RemoteCatalogManager {
...
+                node_id,
+                backend,
+                engine_manager,
+                &catalog_name,
+                &schema_name,
+                schema,
+                &mut max_table_id,
+            )
+            .await?;
+            Ok(max_table_id) as Result<TableId>
```
Comment (WenyXu) [inline src/catalog/src/remote/manager.rs:303]:
```suggestion
            Ok(max_table_id)
```

File: src/catalog/src/remote/manager.rs [hunk 2]
```diff
@@ -332,6 +201,203 @@ impl RemoteCatalogManager {
...
+        }));
+    }
+
+    let join_res = futures::future::join_all(joins).await;
+    for item in join_res {
+        let table_id = item.context(ParallelOpenTableSnafu)??;
+        *max_table_id = (*max_table_id).max(table_id);
+    }
```
Comment (WenyXu) [inline src/catalog/src/remote/manager.rs:311]:
```suggestion
    if let Some(found_max_table_id) = futures::future::try_join_all(joins)
        .await
        .context(ParallelOpenTableSnafu)?
        .into_iter()
        .collect::<Result<Vec<_>>>()?
        .into_iter()
        .max()
    {
        *max_table_id = (*max_table_id).max(found_max_table_id);
    }
```

File: src/catalog/src/remote/manager.rs [hunk 3]
```diff
@@ -332,6 +201,203 @@ impl RemoteCatalogManager {
...
+    Ok(())
+}
+
+/// Iterate over all table entries on metasrv
+async fn iter_remote_tables<'a>(
+    node_id: u64,
+    backend: &'a KvBackendRef,
+    catalog_name: &'a str,
+    schema_name: &'a str,
+) -> Pin<Box<dyn Stream<Item = Result<(TableGlobalKey, TableGlobalValue)>> + Send + 'a>> {
```
Comment (WenyXu) [inline src/catalog/src/remote/manager.rs:322]: Should we consider put `dyn Stream<Item = Result<(TableGlobalKey, TableGlobalValue)>> + Send + 'a>`  to a type, e.g., `pub type XXXType = dyn Stream<Item = Result<(TableGlobalKey, TableGlobalValue)>> + Send + 'a>`

File: src/catalog/src/remote/manager.rs [hunk 4]
```diff
@@ -332,6 +201,203 @@ impl RemoteCatalogManager {
...
+            })
+        })
+        .collect::<Vec<_>>();
+    let vec = futures::future::join_all(joins).await;
+    for res in vec {
+        let table_ref = res.context(ParallelOpenTableSnafu)??;
+        let table_info = table_ref.table_info();
+        let table_name = &table_info.name;
+        let table_id = table_info.ident.table_id;
+        schema.register_table(table_name.clone(), table_ref).await?;
```
Comment (WenyXu) [inline src/catalog/src/remote/manager.rs:383]: Question: should we also register tables in parallel?

Comment (gitccl) [reply src/catalog/src/remote/manager.rs:383]: Is `register_table` thread-safe?

Comment (shuiyisong) [reply src/catalog/src/remote/manager.rs:383]: `register_table` was originally right after `open_or_create_table`, but it used to have `common_runtime::block_on_read`, so it was left serialized.

Now it seems parallelizable to me. Better off, we can try to put it back to near `open_or_create_table`.

cc @v0y4g3r please take a look too

File: src/catalog/src/remote/manager.rs [hunk 5]
```diff
@@ -114,186 +104,65 @@ impl RemoteCatalogManager {
...
-            }
-        }))
-    }
-
     /// Fetch catalogs/schemas/tables from remote catalog manager along with max table id allocated.
     async fn initiate_catalogs(&self) -> Result<HashMap<String, CatalogProviderRef>> {
         let mut res = HashMap::new();
 
         let mut catalogs = self.iter_remote_catalogs().await;
+        let mut joins = Vec::new();
```
Comment (killme2008) [inline src/catalog/src/remote/manager.rs:112]:
```suggestion
        let mut joins = Vec::with_capactity(catalogs.len());
```

File: src/catalog/src/remote/manager.rs (hunk 2 above)
Comment (killme2008) [reply src/catalog/src/remote/manager.rs:311]: Looks like it's not fixed, please close the conversation only when it's fixed.

**Review by killme2008 — COMMENTED** [review]
Great job. Thanks a lot. But I found some CR comments were resolved without actual fixing. Please resolve the comments only when fixing them or leave a comment to explain why not do it.

File: src/catalog/src/remote/manager.rs (hunk 5 above)
Comment (gitccl) [reply src/catalog/src/remote/manager.rs:112]: > Great job. Thanks a lot. But I found some CR comments were resolved without actual fixing. Please resolve the comments only when fixing them or leave a comment to explain why not do it.

Sorry, I fixed it locally, but haven't pushed yet. I will be careful afterwards.

Comment (gitccl) [reply src/catalog/src/remote/manager.rs:112]: It seems that there is no `len` method in Stream.

File: src/catalog/src/remote/manager.rs (hunk 3 above)
Comment (killme2008) [reply src/catalog/src/remote/manager.rs:322]: Is it used in other places? If so, let's do this refactor.

Comment (gitccl) [reply src/catalog/src/remote/manager.rs:322]: It is used only here.

Comment (killme2008) [reply src/catalog/src/remote/manager.rs:322]: Got it. I think this refactor is not necessary. Can you push the latest commit to the remote? And resolve the comments if you already fixed them. Let's keep pushing.

Comment (gitccl) [reply src/catalog/src/remote/manager.rs:319]: Ok

**Review by killme2008 — APPROVED** [review]
LGTM

**Review by WenyXu — APPROVED** [review]

**Review by shuiyisong — APPROVED** [review]
LGTM

**Pull request merged**